### PR TITLE
Fixed logic error of instruction::replace()

### PR DESCRIPTION
**Pull request by tcgu-amd**
Related issue: https://github.com/ROCm/AMDMIGraphX/issues/3110

instruction::replace() performs replace operation on the current instruction then recursively calls itself on any down stream instructions that depends on the output. This is equivalent to performing a pre-order tree walk which is problematic. A simple example would be

op1(op2(op4(x)), op3(op4(x))).

In this example, op1 is a binary operation and op2, op3, op4 are unary operations. The parse tree is simply

  op4
 /   \
op2  op3
 \   /
  op1

An pre-order tree walk on op4.replace will be called in the order of op4 -> op2 -> op1 -> op3

However, at op1, we will encounter an error of shape mismatch because op2 has already been replaced but op3 still hasn't

The correct implementation would be to use BFS, which would yield op4, op2, op3, op1.

### CODE REVIEW

File: src/instruction.cpp [hunk 1]
```diff
@@ -59,14 +59,27 @@ instruction::instruction(literal l)
...
 {
-    if(r != result)
-    {
+    if (r != result){
         result = r;
-        for(auto&& ins : output)
+        std::deque<instruction_ref> q;
+        for (instruction_ref ins: output){
+            q.push_back(ins);
+        }
```
Comment (pfultz2) [inline src/instruction.cpp:67]: Initialize the `deque` with the `output`:  `std::deque<instruction_ref> q(output.begin(), output.end())`.

File: src/instruction.cpp [hunk 2]
```diff
@@ -59,14 +59,27 @@ instruction::instruction(literal l)
...
-            ins->recompute_shape();
-        }
+            shape new_r = compute_shape(ins->op, ins->arguments, ins->module_args);
+            if (new_r != ins->result)
+            {
+                ins->result = new_r;
+                for(instruction_ref child_ins: ins->output)
+                {
+                    q.push_back(child_ins);
+                }
```
Comment (pfultz2) [inline src/instruction.cpp:80]: Use `std::copy` here: `std::copy(ins->output.begin(), ins->output.end(), std::back_inserter(q))`.

File: src/include/migraphx/instruction.hpp [hunk 3]
```diff
@@ -33,6 +33,7 @@
 #include <migraphx/config.hpp>
 #include <string>
 #include <utility>
+#include <deque>
```
Comment (pfultz2) [inline src/include/migraphx/instruction.hpp:36]: Put the include in the .cpp file.

**Comment by pfultz2** [comment]
Unit tests need to be added for this.

File: src/instruction.cpp (hunk 1 above)
Comment (tcgu-amd) [reply src/instruction.cpp]: Fixed

File: src/instruction.cpp (hunk 2 above)
Comment (tcgu-amd) [reply src/instruction.cpp]: Fixed

File: src/include/migraphx/instruction.hpp (hunk 3 above)
Comment (tcgu-amd) [reply src/include/migraphx/instruction.hpp]: Fixed

**Comment by tcgu-amd** [comment]
> Unit tests need to be added for this.

@pfultz2  Should they be included in the scope of this PR? Thanks!

**Comment by pfultz2** [comment]
> Should they be included in the scope of this PR?

Yes, unit tests need to be added to verify the PR works as expected(see [Contributing](https://github.com/ROCm/AMDMIGraphX/wiki/Developer's-Guide#contributing) section). A test case can be added under test/instruction.cpp file.

**Comment by pfultz2** [comment]
Thanks for the test! Just need to fix the licensing and cppcheck issue.

**Pull request merged**